### PR TITLE
allow eviction of DEFAULT template cache - this leaks memory, JVM get OOM fast if the expressions are big.

### PR DESCRIPTION
**Pull request by s50600822**
TemplateFactory.cache leaks memory, JVM get OOM fast if the expressions are big.
Provide methods to clearCache optionally.

### CODE REVIEW

**Comment by Shredder121** [comment]
:+1:
I'm not opposed to the idea.
However could you explain a little on how the cache could keep growing?
Maybe there's something else we also need to address first?

**Comment by karahatherly** [comment]
> However could you explain a little on how the cache could keep growing?

`TemplateFactory.cache` is a `ConcurrentHashMap` so its size is bound only by the number of cache misses. On a project which uses querydsl like Jira does this can result in OOMEs as @s50600822 states. Exposing methods to clear it allows us to

> Maybe there's something else we also need to address first?

I think additionally we should change the `ConcurrentHashMap` to a guava cache and configure it at run-time based on a `CacheBuilderSpec` representation in a property, to allow querydsl consumers to tune that cache if necessary. We'd then still want to expose the `invalidate*` and `cleanUp` methods from `Cache`, similar to what's done in this PR for the `Map`.

**Comment by timowest** [comment]
I'd also like to see a use case description first, how this cache could grow too big. There should be finite amount of operation serialization templates in a project.

**Comment by s50600822** [comment]
It was just a matter of scale. Apparently there is a platform that heavily relies on user inputs and context to build queries to database. 
Adopting serverless architecture of that platform forces the JVMs to serve request from all different customers with different contexts since there's no sticky session.
(that wasn't my idea btw :< ).

**Comment by Shredder121** [comment]
Could you elaborate on the scale? Is it arbitrary amounts of applications in one JVM?

**Comment by dodolzg** [comment]
I have the same problem, java.lang.OutOfMemoryError.
JVM shows that there are lots of char[], string which refrence sql template, however it can be reached by GC ROOT.

**Comment by s50600822** [comment]
> Could you elaborate on the scale? Is it arbitrary amounts of applications in one JVM?

put it this way if you can't think of how big the mem leak is, how about a limit it will never use up says 20MB or something ? The code that generate that many query String into the cache probably need to be fixed, but the library could also help limiting the blast radious.

**Comment by s50600822** [comment]
@Shredder121 do you want to merge it, I don't have write access.

**Comment by jwgmeligmeyling** [comment]
Wouldn't a `Collections.synchronizedMap(new WeakHashMap<String, Template>());` be more appropiate here?

It seems this cache is mostly here to reuse frequently used `Template` instances among operators. Because these are mostly defined in singleton instances (Templates), these are constantly in memory anyway.

Some voices are raising to slowly migrate off from Guava: #2324

**Review by jwgmeligmeyling — CHANGES_REQUESTED** [review]
Please don't merge this yet, I am working on this patch for the 5.x release.

**Pull request closed without merging**
